Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs

```diff
@@ -9,89 +9,140 @@
+    }
+
+    /// <summary>
+    /// Returns the transcoding result from the first matching HttpRule, or
+    /// null if no rules match.
+    /// </summary>
+    internal TranscodingOutput Transcode(IMessage request) =>
+        _transcoders
+            .Select(t => t.Transcode(request))
+            .FirstOrDefault(result => result is not null);
+
+    /// <summary>
+    /// A transcoder for a single rule, ignoring additional bindings.
+    /// </summary>
+    private class SingleRuleTranscoder
+    {
+        private static readonly FieldType[] TypesIneligibleForQueryStringEncoding = { FieldType.Group, FieldType.Message, FieldType.Bytes };
+        private static readonly HttpMethod s_patchMethod = new HttpMethod("PATCH");
+
+        private readonly HttpMethod _httpMethod;
+        private readonly Func<IMessage, string> _bodyTranscoder;
+        private readonly HttpRulePathPattern _pathPattern;
+        private readonly List<FieldDescriptor> _queryParameterFields;
+
+        internal SingleRuleTranscoder(string methodName, MessageDescriptor requestMessage, HttpRule rule)
         {
-            // TODO: The body shouldn't contain any fields that are present in the URI.
-            "*" => (true, null, new Func<IMessage, string>(protoRequest => protoRequest.ToString())),
-            "" => (false, null, new Func<IMessage, string>(protoRequest => null)),
-            string name when requestMessageDescriptor.FindFieldByName(name) is FieldDescriptor field => (false, name, new Func<IMessage, string>(protoRequest => field.Accessor.GetValue(protoRequest).ToString())),
-            _ => throw new ArgumentException($"Method {methodName} has a body parameter {rule.Body} in the 'google.api.http' annotation which is not a field in {requestMessageDescriptor.Name}")
-        };
-
-        _queryParameterFields = new List<FieldDescriptor>();
-        if (!allFieldsInBody)
+            (string pattern, _httpMethod) = rule.PatternCase switch
+            {
+                HttpRule.PatternOneofCase.Get => (rule.Get, HttpMethod.Get),
+                HttpRule.PatternOneofCase.Post => (rule.Post, HttpMethod.Post),
+                HttpRule.PatternOneofCase.Patch => (rule.Patch, s_patchMethod),
+                HttpRule.PatternOneofCase.Put => (rule.Put, HttpMethod.Put),
+                HttpRule.PatternOneofCase.Delete => (rule.Delete, HttpMethod.Delete),
+                HttpRule.PatternOneofCase.Custom => (rule.Custom.Path, new HttpMethod(rule.Custom.Kind)),
+                _ => throw new ArgumentException("HTTP rule has no pattern")
+            };
+
+            _pathPattern = HttpRulePathPattern.Parse(pattern, requestMessage);
+
+            (bool allFieldsInBody, string bodyName, _bodyTranscoder) = rule.Body switch
+            {
+                // TODO: The body shouldn't contain any fields that are present in the URI.
```

Comment:
What I understand is that the body must not be the same as a URI parameters vs. contain.

I also wonder if it's us that need to check for this. We don't check for many of the other constraints specified in the AIP (as far as I can see, but I might be missing something). For instance, we don't check for:

- The body clause must be identical in the top-level annotation and each additional binding.
- RPCs must not define a body at all for RPCs that use the GET or DELETE HTTP verbs.

If anything these should be presubmit checks on protos?

I also understand that, if we don't have those presubmit checks, and we end up executing requests with bad transcoding, errors that end users get will be confusing. In turn, if we check these constraints ourselves we can at least give meaningful errors. But I guess we either check all or none?

Again, I might be missing something here.